File: Day 06/Python/sol.py
"""
FILE: Day 06/Python/sol.py
Author: Alex Jones
Desc: Solution to day 6 problems (11 & 12) for Advent of Code 2024, solved in Python 3.
"""
data = open("Day 06/data.txt", "r").read().strip().replace("\r","").split("\n")

turn_right = {(-1,0): (0,1), (0,1): (1,0), (1,0): (0,-1), (0,-1): (-1,0)}
grid_map = {(i, j): char for i, row in enumerate(data) for j, char in enumerate(row)}
pos = [pos for pos in grid_map if grid_map[pos] == '^'][0]

def simulate(grid_map, pos, dir_, visited):
    seen = set(visited)
    while pos in grid_map and (pos, dir_) not in seen:
        # Maintain a `visited` path to allow shorter simulations in part 2
        visited.append((pos, dir_))
        seen.add((pos, dir_))
        next_pos = (pos[0] + dir_[0], pos[1] + dir_[1])
        if grid_map.get(next_pos) == "#":
            dir_ = turn_right[dir_]
            continue
        pos = next_pos
    return (pos, dir_) not in seen

path = []
simulate(grid_map, pos, (-1,0), path)
print("Problem 11:", len(set(p for (p, _) in path)))

from functools import cache

@cache
def find_next_collision(pos, dir_):
    while pos in grid_map and grid_map[pos] != '#':
        pos = (pos[0] + dir_[0], pos[1] + dir_[1])
    return pos

def between(item, one, two):
    return one <= item < two or two <= item < one

def new_obstacle_blocks_path(current_pos, current_dir, new_pos, obstacle):
    if current_dir[0] != 0 and current_pos[1] == obstacle[1]:
        return between(obstacle[0], current_pos[0], new_pos[0])
    elif current_dir[1] != 0 and current_pos[0] == obstacle[0]:
        return between(obstacle[1], current_pos[1], new_pos[1])
    return False

ans = 0
visited_on_path = set()
# Optimisation 1: Only try placing obstacles on nodes explored in our original path
for i, ((prev_pos, prev_dir), (pos, _)) in enumerate(zip(path[:-1], path[1:])):
    # Only check each position once, the first time encountered
    if pos == prev_pos or pos in visited_on_path: continue
    visited_on_path.add(pos)
    # Optimisation 2: Simulate restoring from just before the point of collision
    current_pos, next_pos, current_dir = prev_pos, prev_pos, prev_dir
    seen = set()
    while next_pos in grid_map and (current_pos, current_dir) not in seen:
        seen.add((current_pos, current_dir))
        # Optimisation 3: Use a memoized global result to automatically find the next
        # collision given the current position and direction. Then, just check if the
        # newly inserted obstacle would have been on that collision path - if so, set
        # the new position to be just before the obstacle instead.
        next_pos = find_next_collision(current_pos, current_dir)
        if new_obstacle_blocks_path(current_pos, current_dir, next_pos, pos):
            next_pos = pos
        current_pos = (next_pos[0] - current_dir[0], next_pos[1] - current_dir[1])
        current_dir = turn_right[current_dir]
    # Only increment if we stopped searching because we found a cycle, instead of 
    # finding the grid border
    ans += next_pos in grid_map

print("Problem 12:", ans)

File: Day 06/Rust/sol.rs
/*
 * FILE: Day 06/Rust/sol.rs
 * Author: Alex Jones
 * Desc: Solution to day 6 problems (11 & 12) for Advent of Code 2024, solved in Rust.
 */
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs;

use itertools::{enumerate, Itertools};

fn turn_right(dir: (i32, i32)) -> (i32, i32) {
    match dir {
        (-1, 0) => (0, 1),
        (0, 1) => (1, 0),
        (1, 0) => (0, -1),
        (0, -1) => (-1, 0),
        other => other,
    }
}

pub fn pos_in_bounds(dims: (usize, usize), pos: (i32, i32)) -> bool {
    0 <= pos.0 && 0 <= pos.1 && (dims.0 as i32) > pos.0 && (dims.1 as i32) > pos.1
}

type State = ((i32, i32), (i32, i32));

pub fn simulate(
    dims: (usize, usize),
    obstacles: &HashSet<(usize, usize)>,
    mut pos: (i32, i32),
    mut dir: (i32, i32),
    visited: &mut Vec<State>,
) -> bool {
    let mut seen = HashSet::<State>::from_iter(visited.iter().cloned());
    while pos_in_bounds(dims, pos) && !seen.contains(&(pos, dir)) {
        // Maintain a `visited` path to allow shorter simulations in part 2
        visited.push((pos, dir));
        seen.insert((pos, dir));
        let next_pos = (pos.0 + dir.0, pos.1 + dir.1);
        if obstacles.contains(&((next_pos.0 as usize), (next_pos.1 as usize))) {
            dir = turn_right(dir);
            continue;
        }
        pos = next_pos;
    }
    !seen.contains(&(pos, dir))
}

type Memo = HashMap<State, (i32, i32)>;
fn find_next_collision(
    memo: &mut Memo,
    dims: (usize, usize),
    obstacles: &HashSet<(usize, usize)>,
    mut pos: (i32, i32),
    dir: (i32, i32),
) -> (i32, i32) {
    let entry = memo.entry((pos, dir));
    if let Entry::Occupied(e) = entry {
        *e.get()
    } else {
        while pos_in_bounds(dims, pos) && !obstacles.contains(&(pos.0 as usize, pos.1 as usize)) {
            pos = (pos.0 + dir.0, pos.1 + dir.1);
        }
        entry.or_insert(pos);
        pos
    }
}

fn between(item: i32, one: i32, two: i32) -> bool {
    one <= item && item < two || two <= item && item < one
}

fn new_obstacle_blocks_path(
    current_pos: (i32, i32),
    current_dir: (i32, i32),
    new_pos: (i32, i32),
    obstacle: (i32, i32),
) -> bool {
    if current_dir.0 != 0 && current_pos.1 == obstacle.1 {
        between(obstacle.0, current_pos.0, new_pos.0)
    } else if current_dir.1 != 0 && current_pos.0 == obstacle.0 {
        between(obstacle.1, current_pos.1, new_pos.1)
    } else {
        false
    }
}

pub fn solve() {
    let contents: String = fs::read_to_string("Day 06/data.txt")
        .expect("Could not read data file")
        .trim()
        .replace("\r", "");
    let data: Vec<&str> = contents.split("\n").collect();

    let mut obstacles: HashSet<(usize, usize)> = HashSet::new();
    let mut guard: Option<(usize, usize)> = None;
    for (i, row) in enumerate(&data) {
        for (j, col) in enumerate(row.bytes()) {
            match col {
                b'^' => guard = Some((i, j)),
                b'#' => drop(obstacles.insert((i, j))),
                _ => (),
            }
        }
    }
    let dims = (data.len(), data[0].len());
    let guard = guard.expect("No guard found?");

    let mut path: Vec<State> = Vec::new();
    assert!(simulate(
        dims,
        &obstacles,
        (guard.0 as i32, guard.1 as i32),
        (-1, 0),
        &mut path
    ));
    println!("Problem 11: {}", path.iter().map(|x| x.0).unique().count());

    let mut memo = Memo::new();

    let mut ans = 0u32;
    let mut visited_on_path = HashSet::new();
    // Optimisation 1: Only try placing obstacles on nodes explored in our original path
    for i in 1..path.len() {
        let pos = path[i].0;
        let (prev_pos, prev_dir) = path[i - 1];
        // Only check each position once, the first time encountered
        if pos == prev_pos || visited_on_path.contains(&pos) {
            continue;
        }
        visited_on_path.insert(pos);
        // Optimisation 2: Restore simulation from just before the point of collision
        let mut current_pos = prev_pos;
        let mut next_pos = prev_pos;
        let mut current_dir = prev_dir;
        let mut seen = HashSet::new();
        while pos_in_bounds(dims, next_pos) && !seen.contains(&(current_pos, current_dir)) {
            seen.insert((current_pos, current_dir));
            // Optimisation 3: Use a memoized global result to automatically find the
            // next collision given the current posiition and direction. Then, just check
            // if the newly inserted obstacle would have been on that collision path - if
            // so, set the new position to be just before the obstacle instead.
            next_pos = find_next_collision(&mut memo, dims, &obstacles, current_pos, current_dir);
            if new_obstacle_blocks_path(current_pos, current_dir, next_pos, pos) {
                next_pos = pos;
            }
            current_pos = (next_pos.0 - current_dir.0, next_pos.1 - current_dir.1);
            current_dir = turn_right(current_dir);
        }
        // Only increment if we stop searching because we found a cycle, instead
        // of finding the grid border.
        ans += pos_in_bounds(dims, next_pos) as u32;
    }
    println!("Problem 12: {}", ans);
}
